kotlin/workflow-ui/core-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt

```diff
@@ -64,28 +66,32 @@ internal class WorkflowRunnerViewModel<OutputT : Any>(
     }
   }
 
-  @Suppress("EXPERIMENTAL_API_USAGE")
-  override val result: Maybe<out OutputT> = session.outputs.asObservable()
-      .firstElement()
-      .doAfterTerminate {
-        scope.cancel(CancellationException("WorkflowRunnerViewModel delivered result"))
-      }
-      .cache()
+  private val result = scope.async {
+    session.outputs.first()
+  }
+
+  override suspend fun awaitResult(): OutputT = result.await()
 
   init {
     @Suppress("EXPERIMENTAL_API_USAGE")
     session.renderingsAndSnapshots
         .map { it.snapshot }
         .onEach { lastSnapshot = it }
         .launchIn(scope)
+
+    // Cancel the entire workflow runtime after the first output is emitted.
+    // Use the Unconfined dispatcher to ensure the cancellation happens as immediately as possible.
+    scope.launch(Dispatchers.Unconfined) {
+      result.join()
+      scope.cancel(CancellationException("WorkflowRunnerViewModel delivered result"))
+    }
```

Comment:
@rjrjr FYI I just reworked this a bit. The result deferred is now created using `async`, which is much more straightforward and means we don't have to manually propagate results and exceptions. The cancel-on-output logic is now responsible _only_ for that, which I think makes it easier to read.